src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -873,19 +873,85 @@ DagState uniqifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<Bo
                             break;
                         case BoundDagTest d:
                             bool foundExplicitNullTest = false;
+                            bool foundExplicitValueTest = false; // TODO: unused for now
                             SplitCases(state, d,
                                 out ImmutableArray<StateForCase> whenTrueDecisions,
                                 out ImmutableArray<StateForCase> whenFalseDecisions,
                                 out ImmutableDictionary<BoundDagTemp, IValueSet> whenTrueValues,
                                 out ImmutableDictionary<BoundDagTemp, IValueSet> whenFalseValues,
-                                ref foundExplicitNullTest);
+                                ref foundExplicitNullTest, ref foundExplicitValueTest);
                             state.TrueBranch = uniqifyState(whenTrueDecisions, whenTrueValues);
                             state.FalseBranch = uniqifyState(whenFalseDecisions, whenFalseValues);
                             if (foundExplicitNullTest && d is BoundDagNonNullTest { IsExplicitTest: false } t)
                             {
                                 // Turn an "implicit" non-null test into an explicit one
                                 state.SelectedTest = new BoundDagNonNullTest(t.Syntax, isExplicitTest: true, t.Input, t.HasErrors);
                             }
+                            else if (d is BoundDagValueTest { IsSubsumptionOnly: true })
+                            {
+                                // A subsumption-only test either gets elided by a preceding source test or we need to handle it here.
```

Comment:
Just to show case the scenario. This is the relevant dag state dump where we were emitting a synthetic test.

```
State 9 REMAINING t1:[2..2147483647] t3:[-2147483648..0] t2:[0..2147483647]
    3. [[<=0, <0, ..] => 1] AND(OR(Not (?Value*(t1 == ValueI32(2: Int32))), Assignment(t4<--t2)), ?Relational(t4 <= ValueDefault(0: Int32)), t5=Indexer(t0[1]), Not (AND(?Value*(t1 == ValueI32(3: Int32)), Assignment(t5<--t2))), OR(Not (?Value*(t1 == ValueI32(2: Int32))), Assignment(t5<--t3)), ?Relational(t5 < ValueDefault(0: Int32)))
    4. [[>0, <=0, ..] => 1] AND(OR(Not (?Value*(t1 == ValueI32(2: Int32))), Assignment(t4<--t2)), ?Relational(t4 > ValueDefault(0: Int32)), t5=Indexer(t0[1]), OR(Not (?Value*(t1 == ValueI32(3: Int32))), Assignment(t5<--t2)), OR(AND(?Value*(t1 == ValueI32(2: Int32)), Assignment(t5<--t3)), ?Relational(t5 <= ValueDefault(0: Int32))))
    5. [[0, 0] => 1] AND(?Value(t1 == ValueI32(2: Int32)), OR(Not (?Value*(t1 == ValueI32(2: Int32))), Assignment(t4<--t2)), ?Value(t4 == ValueDefault(0: Int32)), t5=Indexer(t0[1]), OR(Not (?Value*(t1 == ValueI32(3: Int32))), Assignment(t5<--t2)), OR(Not (?Value*(t1 == ValueI32(2: Int32))), Assignment(t5<--t3)), ?Value(t5 == ValueDefault(0: Int32)))
    6. [{ Length: not 2 }  => 2] Not (?Value(t1 == ValueI32(2: Int32)))
    7. [[var unreachable, var unreachable2] => 3] AND(?Value(t1 == ValueI32(2: Int32)), t5=Indexer(t0[1])) BIND[unreachable=t4; unreachable2=t5]
    Test: ?Value(t1 == ValueI32(2: Int32))
    IsSubsumptionOnly: True
```

<details><summary>Decompiled example with a preceding synthetic tests</summary>

```cs
    // Source
    static int Test(int[] a)
    {
        switch (a)
        {
            case [.., >=0, >0]: return 1;
            case [.., <0, >=0]: return 2;
            case [<=0, <0, ..]: return 3;
            case [>0, <=0, ..]: return 4;
            case [0, 0]: return 5;
            case { Length: not 2 }: return 6;
        }
       return 7;
    }

    // Commit 9
    private static int Test(int[] a)
    {
        if (a != null)
        {
            int length = a.Length;
            if (length >= 2)
            {
                if (a[length - 2] >= 0)
                {
                    if (a[length - 1] > 0)
                    {
                        return 1;
                    }
                    if (a[0] > 0)
                    {
                        int num = a[1];
                        return 4;
                    }
                    if (a[1] >= 0)
                    {
                        if (length != 2)
                        {
                            goto IL_0052;
                        }
                        return 5;
                    }
                }
                else
                {
                    if (a[length - 1] >= 0)
                    {
                        return 2;
                    }
                    // Note: this happens because we have determined that the other indexers are the same
                    // Note: normally we remove these before dag construction but we know it only after which is too late.
                    int num2 = a[0]; 
                    int num3 = a[1];
                }
                return 3;
            }
            goto IL_0052;
        }
        return 7;
        IL_0052:
        return 6;
    }

    // With subsumption checking disabled
    private static int Test(int[] a)
    {
        if (a != null)
        {
            int length = a.Length;
            if (length >= 2)
            {
                int num;
                if (a[length - 2] < 0)
                {
                    if (a[length - 1] < 0)
                    {
                        num = a[0];
                        int num2 = a[1];
                        return 3;
                    }
                    return 2;
                }
                if (a[length - 1] > 0)
                {
                    return 1;
                }
                num = a[0];
                if (num > 0)
                {
                    int num3 = a[1];
                    return 4;
                }
                if (length == 2)
                {
                    if (num != 0 || a[1] != 0)
                    {
                        goto IL_0058;
                    }
                    return 5;
                }
            }
            return 6;
        }
        goto IL_0058;
        IL_0058:
        return 7;
    }
```

Reply:
I am afraid I am not following what are we trying to accomplish in commit 9. It is also not clear to me what conclusions, if any,  could be made based on the information from the previous post in this thread. BTW, it looks like the logic in the code marked "With subsumption checking disabled" doesn't match the original switch. In particular, it looks like it is going to return 3 for  ```[3, 2, -1, -2]```, but that doesn't match expected behavior for the switch.

Reply:
This is WIP, I should've marked it as draft earlier, sorry.

> It is also not clear to me what conclusions, if any, could be made based on the information from the previous post in this thread.

I'm trying to figure out how we could possibly avoid emitting synthetic tests. It makes sense to expect that they wouldn't be really useful in the final code. I am yet to convince myself that it is possible at all (considering how subsumption-checking is implemented to begin with).

Reply:
> BTW, it looks like the logic in the code marked "With subsumption checking disabled" doesn't match the original switch

Updated.

Reply:
>I am not following what are we trying to accomplish in commit 9. 

The idea was to resolve sythetic tests ahead of time. In order to do that, we should still rely on the remaining values on forthcoming tests, that's why it doesn't work.